Drop IsDeprecatedWeakRefSmartPointerException from WebColorChooser

RupinMittal opened the pull request:
#### 24571d6c14c80adfa95c868880c487448d6f6899
<pre>
Drop IsDeprecatedWeakRefSmartPointerException from WebColorChooser
<a href="https://bugs.webkit.org/show_bug.cgi?id=281720">https://bugs.webkit.org/show_bug.cgi?id=281720</a>
<a href="https://rdar.apple.com/138165163">rdar://138165163</a>

Reviewed by Geoffrey Garen and Chris Dumez.

Remove the exception by making WebColorChooser ref-counted. Since it&apos;s now
ref-counted, it can out-live it&apos;s m_colorChooserClient member. So we make
m_colorChooserClient a WeakPtr and null-check before use.

* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::setValue):
(WebCore::ColorInputType::showPicker):
(WebCore::ColorInputType::endColorChooser):
* Source/WebCore/html/ColorInputType.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyChromeClient::createColorChooser):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createColorChooser):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/ColorChooser.h:
* Source/WebCore/platform/ColorChooserClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createColorChooser):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.cpp:
(WebKit::WebColorChooser::didChooseColor):
(WebKit::WebColorChooser::didEndChooser):
(WebKit::WebColorChooser::reattachColorChooser):
* Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.h:
(WebKit::WebColorChooser::create):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::didEndColorPicker):
(WebKit::WebPage::didChooseColor):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createColorChooser):

Canonical link: <a href="https://commits.webkit.org/285496@main">https://commits.webkit.org/285496@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/832823dc39a7596a2578096cca4d2dd2dc992819

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72719 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52144 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76909 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74834 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23816 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57178 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15676 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47155 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62590 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37605 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43804 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20056 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22282 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65657 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20415 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16965 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19597 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65629 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64906 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13215 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6862 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11192 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47942 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2729 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49009 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50304 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -271,7 +271,7 @@ void ColorInputType::showPicker()
             m_chooser = chrome->createColorChooser(*this, valueAsColor());
         else
-            m_chooser->reattachColorChooser(valueAsColor());
+            RefPtr { m_chooser }->reattachColorChooser(valueAsColor());
```
We could also just reverse the `if` condition:
```
if (RefPtr chooser = m_chooser)
    chooser->reattachColorChooser(valueAsColor());
else
    m_chooser = chrome->createColorChooser(*this, valueAsColor());
```

geoffreygaren replied:
I guess this is the case where we've been adding `protectedChooser()`. Or Maciej's suggestion of `m_chooser->protected()`. I'll bring this up on Slack again. But doesn't block your patch.

geoffreygaren reviewed (approved):
r=me

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.h:
```diff
@@ -50,11 +40,19 @@ namespace WebKit {
 
-class WebColorChooser : public WebCore::ColorChooser, public CanMakeWeakPtr<WebColorChooser> {
+class WebColorChooser : public WebCore::ColorChooser, public RefCountedAndCanMakeWeakPtr<WebColorChooser> {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(WebColorChooser);
```
I don't think this needs to be INLINE. We can probably just use `WTF_MAKE_TZONE_ALLOCATED` her and `WTF_MAKE_TZONE_ALLOCATED_IMPL` in the cpp like we usually do.

RupinMittal replied:
I made it INLINE because the abstract base class `ColorChooser` is INLINE. Should I change the base class to `WTF_MAKE_TZONE_ALLOCATED` as well?

cdumez replied:
The base class is probably inline because there is no CPP. As far as I know nothing prevents the base class from being inline and the subclass from being out of line. Therefore, I don't see why you'd need to modify the base class.

RupinMittal replied:
Ok that makes sense. Fixed!

RupinMittal committed:
Drop IsDeprecatedWeakRefSmartPointerException from WebColorChooser
https://bugs.webkit.org/show_bug.cgi?id=281720
rdar://138165163

Reviewed by Geoffrey Garen and Chris Dumez.

Remove the exception by making WebColorChooser ref-counted. Since it's now
ref-counted, it can out-live it's m_colorChooserClient member. So we make
m_colorChooserClient a WeakPtr and null-check before use.

* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::setValue):
(WebCore::ColorInputType::showPicker):
(WebCore::ColorInputType::endColorChooser):
* Source/WebCore/html/ColorInputType.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyChromeClient::createColorChooser):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createColorChooser):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/platform/ColorChooser.h:
* Source/WebCore/platform/ColorChooserClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createColorChooser):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.cpp:
(WebKit::WebColorChooser::didChooseColor):
(WebKit::WebColorChooser::didEndChooser):
(WebKit::WebColorChooser::reattachColorChooser):
* Source/WebKit/WebProcess/WebCoreSupport/WebColorChooser.h:
(WebKit::WebColorChooser::create):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::didEndColorPicker):
(WebKit::WebPage::didChooseColor):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createColorChooser):

Canonical link: https://commits.webkit.org/285496@main
